### PR TITLE
[WIP] Logshipper

### PR DESCRIPTION
**Pull request by triplem**
Add possiblity to send logs to a specific email address, which is fully customizable.

### CODE REVIEW

**Comment by MilosKozak** [comment]
@triplem great. I like it!
Suggestions:
* change Logshipper to Maintenance
* Move there "DB reset", "Export/Import"
* do not lock visible by default
* not sure what with "show log".  It displays only short time range and it's not filtered by app. Not much usable. Maybe replace with displaying last log file?

@AdrianLxM @jotomo

File: app/src/main/java/info/nightscout/androidaps/plugins/LogShipper/LogShipperPlugin.java [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        }
+    }
+
+    /**
+     * returns a list of log files. The number of returned logs is given via the amount
+     * parameter. The log files are sorted by the name descending.
+     *
+     * @param directory
+     * @param amount
+     * @return
```
Comment (jotomo) [inline app/src/main/java/info/nightscout/androidaps/plugins/LogShipper/LogShipperPlugin.java]: Please don't add useless auto-generated docs ...

File: app/src/test/java/info/nightscout/utils/LoggerUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+import org.junit.Test;
+
+import java.io.File;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class LoggerUtilsTest {
```
Comment (jotomo) [inline app/src/test/java/info/nightscout/utils/LoggerUtilsTest.java]: Missing tests here or obsolete class?

File: app/src/main/java/info/nightscout/androidaps/plugins/LogShipper/LogShipperPlugin.java [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        LOG.debug("zipFile: {}", zipFile.getAbsolutePath());
+        File zip = this.zipLogs(zipFile, logs);
+
+        Uri attachementUri = FileProvider.getUriForFile(this.ctx, "info.nightscout.androidaps.fileprovider", zip);
+        Intent emailIntent = this.sendMail(attachementUri, recipient, "Log Export");
+        LOG.debug("sending emailIntent");
+        ctx.startActivity(emailIntent);
+    }
+
+    public void deleteLogs() {
```
Comment (jotomo) [inline app/src/main/java/info/nightscout/androidaps/plugins/LogShipper/LogShipperPlugin.java]: I don't really see the need for the user to initiate deletion of logs. We could automatically remove all logs older than x on boot (or just retain the last x files)?

**Comment by jotomo** [comment]
@MilosKozak I don't really see who'd want to view logs on the phone - users don't understand it and as a dev I open logs on bigger screens.
How about removing "Show logs" from the overflow menu and putting "Email logs" there, so support can just ask users to quickly do that (without having to go into the maintenance plugin or make it visible)

**Comment by MilosKozak** [comment]
yeah show logs is useless. i want to hear opinion of others. and upper right menu is now pretty full too

**Comment by triplem** [comment]
I do like the idea of putting all "Maintenance" things together in one Plugin. Some ideas I have (check updates, ..) do fit into this category as well. 

Show Logs is really rather useless. So I am all for this new Plugin.

File: app/src/main/java/info/nightscout/androidaps/plugins/LogShipper/LogShipperPlugin.java (hunk 3 above)
Comment (triplem) [reply app/src/main/java/info/nightscout/androidaps/plugins/LogShipper/LogShipperPlugin.java]: Well, this is a lot ack config anyways.

File: app/src/main/res/xml/pref_logshipper.xml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        android:key="logshipper"
+        android:title="@string/logshipper_settings">
+
+        <com.andreabaccega.widget.ValidatingEditTextPreference
+            android:defaultValue="logs@androidaps.org"
+            android:dialogMessage="@string/logshipper_email"
+            android:inputType="textEmailAddress"
+            android:key="@string/key_logshipper_email"
+            android:selectAllOnFocus="true"
+            android:title="@string/nsclientinternal_url_title"
```
[the remark this answers is not in the document]
Comment (triplem) [reply app/src/main/res/xml/pref_logshipper.xml]: keys should be IMHO not using the strings, just to be on the safe side? no? rather redundant anyways, that is true.

[the remark this answers is not in the document]
Comment (triplem) [reply app/src/main/res/xml/pref_logshipper.xml]: ah, sorry, you are right, changed already.

File: app/src/main/java/info/nightscout/androidaps/plugins/LogShipper/LogShipperPlugin.java [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+                .pluginName(R.string.logshipper)
+                .shortName(R.string.logship_shortname)
+                .preferencesId(R.xml.pref_logshipper)
+                .description(R.string.description_logship)
+        );
+        this.ctx = ctx;
+    }
+
+    @Override
+    protected void onStart() {
```
[the remark this answers is not in the document]
Comment (triplem) [reply app/src/main/java/info/nightscout/androidaps/plugins/LogShipper/LogShipperPlugin.java]: removed. thanks for spotting.

File: app/src/main/java/info/nightscout/androidaps/plugins/LogShipper/LogShipperPlugin.java (hunk 1 above)
Comment (triplem) [reply app/src/main/java/info/nightscout/androidaps/plugins/LogShipper/LogShipperPlugin.java]: yeah, well, information about sort order is imho rather useful.

File: app/src/test/java/info/nightscout/utils/LoggerUtilsTest.java (hunk 2 above)
Comment (triplem) [reply app/src/test/java/info/nightscout/utils/LoggerUtilsTest.java]: not used anymore, deleted.

**Comment by MilosKozak** [comment]
save context not in `onCreate` but in `onResume`.  in `onPause` invalidate the context

File: app/src/main/java/info/nightscout/androidaps/plugins/LogShipper/LogShipperPlugin.java (hunk 1 above)
Comment (jotomo) [reply app/src/main/java/info/nightscout/androidaps/plugins/LogShipper/LogShipperPlugin.java]: Agreed. I was referring to the _param_ and _return_ tags without a description :-)

**Pull request merged**